public/app/features/dashboard/panel_editor/state/reducers.ts

```diff
@@ -0,0 +1,72 @@
+import { reducerFactory } from '../../../../core/redux';
+import { panelEditorCleanUp, panelEditorInitCompleted, panelEditorChangeTab } from './actions';
+
+export interface PanelEditorTab {
+  id: string;
+  text: string;
+}
+
+export enum PanelEditorTabIds {
+  Queries = 'queries',
+  Visualization = 'visualization',
+  Advanced = 'advanced',
+  Alert = 'alert',
+}
+
+export const panelEditorTabTexts = {
+  [PanelEditorTabIds.Queries]: 'Queries',
+  [PanelEditorTabIds.Visualization]: 'Visualization',
+  [PanelEditorTabIds.Advanced]: 'General',
+  [PanelEditorTabIds.Alert]: 'Alert',
+};
+
+export const getPanelEditorTab = (tabId: PanelEditorTabIds): PanelEditorTab => {
+  return {
+    id: tabId,
+    text: panelEditorTabTexts[tabId],
+  };
+};
+
+export interface PanelEditorState {
+  activeTab: PanelEditorTabIds;
+  tabs: PanelEditorTab[];
+}
+
+export const initialState: PanelEditorState = {
+  activeTab: null,
+  tabs: [],
+};
+
+export const panelEditorReducer = reducerFactory<PanelEditorState>(initialState)
+  .addMapper({
+    filter: panelEditorInitCompleted,
+    mapper: (state, action): PanelEditorState => {
+      const { activeTab, tabs } = action.payload;
+      return {
+        ...state,
+        activeTab,
+        tabs,
+      };
+    },
+  })
+  .addMapper({
+    filter: panelEditorCleanUp,
+    mapper: (): PanelEditorState => initialState,
+  })
+  .addMapper({
+    filter: panelEditorChangeTab,
```

Comment:
Having to keep two separate states in sync now (activeTab from url and the new redux state) is that ideal?  Before it was only the location query param, and now we have two states that that need to be kept in sync. Seeing a bug where if we update query tab via browser back button it does not change tab in Grafana (this seems to be a bug in master as well).

Reply:
Good points I will look into this closer

Reply:
Using only the prop from `location` now so that we can navigate to a tab from outside PanelEditor, again good catch!

Reply:
great!